subsys/fs/nvs/README.md

```diff
@@ -0,0 +1,163 @@
+
+Suppose we use a 4 byte bit state variable that is changed every minute and needs to be restored after reboot. NVS has been defined with a sector_size equal to the pagesize (1024 bytes) and 2 sectors have been defined. Each sector in NVS has a sector header of 8 byte.
+Each write of the state variable requires 4 byte for the variable, but also 4
+byte for the data header and 4 byte for the data slot, so in total 12 bytes.
+When storing the data the first sector will be full after (1024-8)/12 min.
+Then the second sector is used and after another 84 min this is also full. When
+this is the case the first sector should be used for storage and the first
+sector is erased. So this is after 168 min. The expected life of the device is
+168 min x 20000 or about 6.5 years.
+
+With NS the amount of storages/min, DS the data size in bytes and PAGE_ERASES
+the number of times a page can be erased, the expected life in minutes can be
+calculated as:
+```c
+  SECTOR_COUNT*(SECTOR_SIZE(in bytes)-8)*PAGE_ERASES/(NS*(DS(in bytes)+8) min
+```
+From this formula it is also clear what to do in case the expected life is to
+short: increase SECTOR_COUNT or SECTOR_SIZE.
+
+# Sample
+
+A Sample of how NVS can be used is supplied in ```samples/subsys/nvs```.
+
+
+# High level API
+
+In the high level API the following routines are available:
+
+nvs_init()
+  - initialize nvs for a given nvs_fs
+
+nvs_clear()
+  - clear the nvs_fs, deleting all Storage
+
+nvs_write()
+  - write an entry to flash
+
+nvs_delete()
+  - deletes an entry from flash
+
+nvs_read()
+  - read the latest entry from flash
+
+nvs_read_hist()
+  - read older entry from flash. This requires first to read the oldest one: nvs_read_hist(..., 0) and then with consecutive nvs_read_hist(..., 1) all history items can be read.
+
+To avoid frequent erases of flash sectors the maximum size of the entries that can be written in the high level API is limited to SECTOR_SIZE/4.
+
+# Usage of high level API:
+
+1. Initialize the filesystem with nvs_init
+
+2. To add an element to nvs: call nvs_write() to write an entry to flash
+
+3. To read latest (most recent) entry in nvs: call nvs_read() to read the entry from flash
+
+# Low Level API
+
+The low level API gives an interface that is somewhat similar to FCB. The low level API allows some more detailed control and provides the following routines:
+
+nvs_append()
+  - reserve space to store an element - writes header to flash and returns write location
+
+nvs_flash_write()
+  - write contents to flash
+
+nvs_append_close()
+  - calculate crc16 and write to flash
+
+nvs_get_first_entry()
+  - get the first (oldest) entry in flash of specific id: returns read location and length
+
+nvs_get_last_entry()
+  - get the last (newest) entry in flash of specific id: returns read location and length
+
+nvs_walk_entry()
+  - get the next entry in flash of specific id: returns read location and length (entry needs to be initialised with data location)
```

Comment:
next entry from where? from the beginning?

Reply:
Next entry from the supplied entry.